irmaseal-core/src/stream/opener.rs

```diff
@@ -1,64 +1,64 @@
-pub struct OpenerUnsealed<R: Readable> {
-    aes: SymCrypt,
-    hmac: Verifier,
+/// Enables the library user to lookup the UserSecretKey corresponding to this Identity
+/// and offers the unseal method to start decrypting the encrypted content of the bytestream.
+pub struct OpenerSealed<R: AsyncRead + Unpin> {
     r: R,
-    resultbuf: Option<[u8; BLOCKSIZE]>,
+    metadata_buf: ArrayVec<[u8; MAX_METADATA_SIZE]>,
 }
 
-impl<R: Readable> OpenerSealed<R> {
+impl<R: AsyncRead + Unpin> OpenerSealed<R> {
     /// Starts interpreting a bytestream as an IRMAseal stream.
     /// Will immediately detect whether the bytestream actually is such a stream, and will yield
     /// the identity for which the stream is intended, as well as the stream continuation.
-    pub fn new(r: R) -> Result<(Metadata, OpenerSealed<R>), Error> {
-        let mut ar = ArchiveReader::<R, [u8; MAX_METADATA_SIZE]>::new(r);
+    pub async fn new(mut r: R) -> Result<(Metadata, Self), Error> {
+        let mut buffer = [0u8; 4 + core::mem::size_of::<u16>()];
+        r.read_exact(&mut buffer)
+            .map_err(|err| Error::ReadError(err))
+            .await?;
 
-        let prelude = ar.read_bytes_strict(PRELUDE.len())?;
-        if prelude != PRELUDE {
+        if buffer[..4] != PRELUDE {
             return Err(Error::NotIRMASEAL);
         }
 
-        let meta_len = u16::from_be_bytes(
-            ar.read_bytes_strict(core::mem::size_of::<u16>())?
-                .try_into()
-                .unwrap(),
-        );
-        if usize::from(meta_len) > MAX_METADATA_SIZE {
+        let meta_len = usize::from(u16::from_be_bytes(buffer[4..].try_into().unwrap()));
+        if meta_len > MAX_METADATA_SIZE {
             return Err(Error::FormatViolation);
         }
 
-        let metadata_buf = ar.read_bytes_strict(meta_len.into())?;
+        let mut metadata_buf: ArrayVec<[u8; MAX_METADATA_SIZE]> = ArrayVec::new();
+        unsafe {
+            // set_len can only fail when the length is greater than MAX_METADATA_SIZE
+            // and we explicitly check on this case above, so this cannot happen.
+            metadata_buf.set_len(meta_len);
+        }
+        r.read_exact(metadata_buf.as_mut_slice())
+            .map_err(|err| Error::ReadError(err))
+            .await?;
 
-        let metadata = postcard::from_bytes(metadata_buf).or(Err(Error::FormatViolation))?;
+        let metadata =
+            postcard::from_bytes(metadata_buf.as_slice()).or(Err(Error::FormatViolation))?;
 
-        Ok((metadata, OpenerSealed { ar }))
+        Ok((metadata, OpenerSealed { r, metadata_buf }))
     }
 
-    /// Will unseal the stream continuation and yield a plaintext bytestream.
-    pub fn unseal(
-        self,
+    /// Will unseal the stream continuation and write the plaintext in the given writer.
+    pub async fn unseal<W: AsyncWrite + Unpin>(
+        mut self,
         metadata: &Metadata,
```

Comment:
Isn't it a bit weird that `new` returns the `Metadata` and `unseal` exactly gets it again as parameter? Wouldn't it be more logical to store it in the `OpenerSealed` struct and return a pointer in `new` or make a getter to get it from the struct?